src/software/ai/hl/stp/play/free_kick/free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,317 @@
+#include "free_kick_play_fsm.h"
+
+FreeKickPlayFSM::FreeKickPlayFSM(const TbotsProto::AiConfig &ai_config)
+    : ai_config(ai_config),
+      align_to_ball_tactic(std::make_shared<MoveTactic>()),
+      shoot_tactic(std::make_shared<KickTactic>()),
+      chip_tactic(std::make_shared<ChipTactic>()),
+      passer_tactic(std::make_shared<KickTactic>()),
+      receiver_tactic(
+          std::make_shared<ReceiverTactic>(ai_config.receiver_tactic_config())),
+      receiver_positioning_tactics(
+          {std::make_shared<MoveTactic>(), std::make_shared<MoveTactic>()}),
+      defense_play(std::make_shared<DefensePlay>(ai_config)),
+      receiver_position_generator(ReceiverPositionGenerator<EighteenZoneId>(
+          std::make_shared<const EighteenZonePitchDivision>(
+              Field::createSSLDivisionBField()),
+          ai_config.passing_config())),
+      pass_generator(ai_config.passing_config()),
+      best_pass_and_score_so_far(
+          PassWithRating{.pass = Pass(Point(), Point(), 0), .rating = 0})
+{
+}
+
+void FreeKickPlayFSM::setupPosition(const Update &event)
+{
+    PriorityTacticVector tactics_to_run = {{}};
+
+    // Passing robot
+    updateAlignToBallTactic(event.common.world_ptr);
+    tactics_to_run[0].emplace_back(align_to_ball_tactic);
+
+    // Already assigned one tactic to align to ball.
+    // Assign up to 2 receivers and the remaining tactics are defenders
+    setReceiverAndDefenderTactics(tactics_to_run, event, 2, 1);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+bool FreeKickPlayFSM::setupDone(const Update &event)
+{
+    return align_to_ball_tactic->done();
+}
+
+void FreeKickPlayFSM::updateReceiverPositioningTactics(
+    const WorldPtr world, unsigned int num_tactics,
+    const std::vector<Point> &existing_receiver_positions,
+    const std::optional<Point> &pass_origin_override)
+{
+    // These two tactics will set robots to roam around the field, trying to put
+    // themselves into a good position to receive a pass
+    if (num_tactics != receiver_positioning_tactics.size())
+    {
+        receiver_positioning_tactics =
+            std::vector<std::shared_ptr<MoveTactic>>(num_tactics);
+        std::generate(receiver_positioning_tactics.begin(),
+                      receiver_positioning_tactics.end(),
+                      []() { return std::make_shared<MoveTactic>(); });
+    }
+
+    std::vector<Point> best_receiving_positions =
+        receiver_position_generator.getBestReceivingPositions(
+            *world, num_tactics, existing_receiver_positions, pass_origin_override);
+    // Note that getBestReceivingPositions may return fewer positions than requested
+    // if there are not enough robots, so we will need to check the size of the vector.
```

Comment:
isn't it guaranteed to be exactly num_tactics? since `receiver_positioning_tactics` is resized to `num_tactics`

Reply:
We actually update `num_tactics` to be a valid number within `getBestReceivingPositions`.
https://github.com/UBC-Thunderbots/Software/blob/2b916894500433799bb35e5ec4a074d5ea367fc5/src/software/ai/passing/receiver_position_generator.hpp#L141-L152